### ISSUE TITLE
eDiff-I: Text-to-Image Diffusion Models with an Ensemble of Expert Denoisers

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#189]
# TL; DR

- Train an ensemble of expert denoisers to leverage the temporal dynamics of generative diffusion process.
- Use T5 Text, CLIP Text, and CLIP Image encoder to strengthen and extend the conditioning mechanism.

# Motivation

- Prior text-to-image diffusion models used a single model for the whole denoising pipeline (high to low noise level), which limits its capacity to model textual information with high fidelity.
- Prior models used a single text encoder, typically T5 or CLIP to provide conditioning information to the synthesis framework.

# Formulation

### Key Observation

![image](https://user-images.githubusercontent.com/46308407/202398453-90da55a1-e118-4607-b960-f9cd74c817cd.png)
![image](https://user-images.githubusercontent.com/46308407/202398488-9d74ddb9-9d3e-4e8c-8d2b-16ac4f3c4765.png)

- In their empirical observations, the textual information plays a critical role in the early denoising process (with high noise level) but the model tends to ignore this information in the later stage, as the coarse layout of the image has already been determined.

### Ensemble of Experts

![image](https://user-images.githubusercontent.com/46308407/202397922-6ecb7766-1af3-4a94-a2a5-7cf1f57c80c7.png)

- The core idea is to train an ensemble of expert denoisers that are responsible for different stages of the denoising pipeline. 
- To achieve a good trade-off between sample quality and computational cost, they decide to train atmost 3 experts, one responsible for the beginning, another for the end, and the last one for the intermediate steps. (upon recursive binary split of the horizontal denoising process)

### Multiple Conditioning

![image](https://user-images.githubusercontent.com/46308407/202398849-44305524-42ab-4871-9fbb-ab743e587741.png)

- On top of cascaded diffusion model pipeline, they test providing multiple conditioning embeddings to the diffusion model, namely by using T5 text encoder, CLIP text encoder and CLIP image encoder.
- For the text input, using T5 and CLIP together gave the best output, as CLIP does well on composing the global structure of the image while T5 excels at processing long and complex text queries.
- By using CLIP image embeddings as additional inputs, eDiff-I can perform novel applications like style transfer (using the reference image embedding).

### Paint-with-words

![image](https://user-images.githubusercontent.com/46308407/202399316-e6f20fd5-30f7-41ab-a47c-dedee6305844.png)

- The application is similar to SPADE (GAUGAN) and the mechanism is (likely, in my opinion) inspired by prompt-to-prompt (#181).
- The user can draw coarse masks by selecting each phrase as the *brush*, and this can be used to form the text-image cross-attention map to guide the structure of the synthesized image.
- Formally, they add this user-specified spatial mask to the QK cross-attention map between text condition and image patches, where a scalar value w is used to control the relative strength of this signal.

# Evaluation

### Quantitative

![image](https://user-images.githubusercontent.com/46308407/202399938-5e30a5ae-abe0-4899-8ea6-7d45fd28396e.png)

- Ensemble of experts helps, as well as ensemble of text encoders.

![image](https://user-images.githubusercontent.com/46308407/202400082-f096f7b0-59f7-487b-853f-fae39bcbf8e9.png)

### Qualitative

![image](https://user-images.githubusercontent.com/46308407/202400148-911fca3b-c5b3-429f-b39c-b27976787ff8.png)

- Again, ensemble helps.

![image](https://user-images.githubusercontent.com/46308407/202400214-eac332a1-cf60-47c7-969d-6342f3769860.png)
![image](https://user-images.githubusercontent.com/46308407/202400231-cf34ca7a-d353-41ff-85d2-cd15cc8fcceb.png)
![image](https://user-images.githubusercontent.com/46308407/202400375-54be81df-1d98-4c29-b3d5-7caf2657edc0.png)

- eDiff-I outperforms powerful baselines on challenging settings like multiple entity, text within the image and long-tailed caption.

![image](https://user-images.githubusercontent.com/46308407/202400457-db2beb22-de51-4b43-ac20-74056c8fa81c.png)

![image](https://user-images.githubusercontent.com/46308407/202400482-77be66d7-548f-46b7-8f14-fe026a7b99f1.png)

### Using CLIP Image Embeddings

![image](https://user-images.githubusercontent.com/46308407/202400512-d89c38e8-773c-4d64-a8b3-672ff212bb37.png)

### Paint-with-words

![image](https://user-images.githubusercontent.com/46308407/202400595-b5c82ac3-7fe1-44c3-832d-b2776549f3c2.png)

# Inspiration

- There is a great deal of ongoing research on the working dynamics of Diffusion Models. This paper adds insights to it by analyzing how the relative importance of the conditioning information changes over the course of denoising.
- In case of cross-attention-based conditioning, direct operation on the cross-attention map seems to work well, especially with diffusion models. Similar insights from Prompt-to-Prompt.
- The fact that we can additionally condition our synthesis on image data is intriguing. With a competent joint embedding model, other modalities may further be used.
